File: Producto/frontend/src/remote/DTO/EstadoCosteo.js
/**
 * Espejo del enum EstadoCosteo del backend.
 *
 * Ciclo de vida:
 *   BORRADOR  → COSTEADO, RECHAZADO
 *   COSTEADO  → APROBADO, RECHAZADO, BORRADOR (reproceso)
 *   APROBADO  → (terminal)
 *   RECHAZADO → BORRADOR (reproceso)
 */
export const EstadoCosteo = Object.freeze({
    BORRADOR:  'BORRADOR',
    COSTEADO:  'COSTEADO',
    APROBADO:  'APROBADO',
    RECHAZADO: 'RECHAZADO',
});

/** Etiqueta legible para mostrar en la UI. */
export const ESTADO_COSTEO_LABEL = Object.freeze({
    [EstadoCosteo.BORRADOR]:  'Borrador',
    [EstadoCosteo.COSTEADO]:  'Costeado',
    [EstadoCosteo.APROBADO]:  'Aprobado',
    [EstadoCosteo.RECHAZADO]: 'Rechazado',
});

/**
 * Devuelve true si la transición origen → destino es válida,
 * replicando la lógica del backend `EstadoCosteo.puedeTransicionarA`.
 */
export function puedeTransicionarA(origen, destino) {
    if (!origen || !destino) return false;
    switch (origen) {
        case EstadoCosteo.BORRADOR:
            return destino === EstadoCosteo.COSTEADO || destino === EstadoCosteo.RECHAZADO;
        case EstadoCosteo.COSTEADO:
            return destino === EstadoCosteo.APROBADO || destino === EstadoCosteo.RECHAZADO || destino === EstadoCosteo.BORRADOR;
        case EstadoCosteo.APROBADO:
            return false;
        case EstadoCosteo.RECHAZADO:
            return destino === EstadoCosteo.BORRADOR;
        default:
            return false;
    }
}

export default EstadoCosteo;

File: Producto/backend/src/main/java/backend/com/produccion/domain/enums/EstadoCosteo.java
package backend.com.produccion.domain.enums;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

/**
 * Ciclo de vida del Costeo.
 *
 * <pre>
 * BORRADOR  → COSTEADO, RECHAZADO
 * COSTEADO  → APROBADO, RECHAZADO, BORRADOR (reproceso)
 * APROBADO  → (terminal)
 * RECHAZADO → BORRADOR (reproceso)
 * </pre>
 *
 * - BORRADOR: estado inicial (pre-costeo creado al emitir la SCOS).
 * - COSTEADO: producción incorporó/confirmó los costos.
 * - RECHAZADO: hubo diferencias; exige motivo y versiona el costeo rechazado.
 * - APROBADO: habilita vincular el costeo a ítems de EVN/NV.
 */
@Getter
@RequiredArgsConstructor
public enum EstadoCosteo {
    BORRADOR("Borrador"),
    COSTEADO("Costeado"),
    APROBADO("Aprobado"),
    RECHAZADO("Rechazado");

    private final String descripcion;

    /** Indica si desde este estado es válido transicionar al estado destino. */
    public boolean puedeTransicionarA(EstadoCosteo destino) {
        if (destino == null) return false;
        return switch (this) {
            case BORRADOR  -> destino == COSTEADO || destino == RECHAZADO;
            case COSTEADO  -> destino == APROBADO || destino == RECHAZADO || destino == BORRADOR;
            case APROBADO  -> false;
            case RECHAZADO -> destino == BORRADOR;
        };
    }
}
